tests/net/test_crawler.py

```diff
@@ -72,3 +74,291 @@ def test_extract_disconnect_urls_multiple_urls():
+    respx.get(target_url).mock(
+        return_value=httpx.Response(
+            302,
+            headers={"Location": "http://perdu.com/2"},
+        )
+    )
+
+    respx.get("http://perdu.com/2").mock(
+        return_value=httpx.Response(
+            302,
+            headers={"Location": "http://perdu.com/3"},
+        )
+    )
+
+    respx.get("http://perdu.com/3").mock(
+        return_value=httpx.Response(
+            302,
+            headers={"Location": "http://perdu.com/4"},
+        )
+    )
+
+    respx.get("http://perdu.com/4").mock(
+        return_value=httpx.Response(
+            302,
+            headers={"Location": "http://perdu.com/5"},
+        )
+    )
+
+    respx.get("http://perdu.com/5").mock(
+        return_value=httpx.Response(
+            200,
+            text="OK"
+        )
+    )
+
+    crawler = AsyncCrawler(target_url, timeout=1)
+
+    page = await crawler.async_get(Request(target_url), follow_redirects=True)
+    assert page.status == 200
+    assert page.content == "OK"
+
+
+@respx.mock
+@pytest.mark.asyncio
+async def test_async_try_login_basic_digest_ntlm_wrong_credentials():
+    target_url = "http://perdu.com/"
+    respx.get(target_url).mock(
+        return_value=httpx.Response(
+            200,
+            text="<html><head><title>Vous Etes Perdu ?</title></head><body><h1>Perdu sur l'Internet ?</h1> \
+            <h2>Pas de panique, on va vous aider</h2> \
+            <strong><pre>    * <----- vous &ecirc;tes ici</pre></strong><a href='http://perdu.com/foobar/'></a> \
+            <a href='http://perdu.com/foobar/signout'></a> \
+            <div><a href='http://perdu.com/a/b/signout'></a></div></body></html>"
+        )
+    )
+
+    auth_url_1 = "http://perdu.com/login1"
+    respx.get(auth_url_1).mock(
+        return_value=httpx.Response(
+            401,
+            text="KO"
+        )
+    )
+
+    auth_url_2 = "http://perdu.com/login2"
+    respx.get(auth_url_2).mock(
+        return_value=httpx.Response(
+            403,
+            text="KO"
+        )
+    )
+
+    auth_url_3 = "http://perdu.com/login3"
+    respx.get(auth_url_3).mock(
+        return_value=httpx.Response(
+            404,
+            text="KO"
+        )
+    )
+
+    crawler = AsyncCrawler(target_url, timeout=1)
+    crawler._auth_credentials = ["username", "password"]
+
+    is_logged_in, form, disconnect_urls = await crawler._async_try_login_basic_digest_ntlm(auth_url_1)
+
+    assert is_logged_in is False
+    assert len(form) == 0
+    assert len(disconnect_urls) == 0
+
+    is_logged_in, form, disconnect_urls = await crawler._async_try_login_basic_digest_ntlm(auth_url_2)
+
+    assert is_logged_in is False
+    assert len(form) == 0
+    assert len(disconnect_urls) == 0
+
+    is_logged_in, form, disconnect_urls = await crawler._async_try_login_basic_digest_ntlm(auth_url_3)
+
+    assert is_logged_in is False
+    assert len(form) == 0
+    assert len(disconnect_urls) == 0
```

Comment:
You can use a for loop here

Reply:
In fact, you can use a for loop for the whole test

Reply:
Done :rocket: